Add timezone correction to the Coming Episodes Page

Prinz23 opened the pull request:
It's automatically corrected via the dateutil lib. 

The local timezone is detected and the show timezone is detected from the network. The network timezone table is loaded and updated with every version check from https://github.com/Prinz23/sb_network_timezones

Also the zoneinfo of the dateutil is also automatically updated from that url if needed.

And I added the the possibility to set the missing days shown on the Coming Episodes Page in the Config -> Generals Menu. 2 day to 30 days allowed.

I added runtime import from the tvdb, because a show will be marked as missed after airtime + runtime is over. 

All int convert functions are replaced with a new helpers function helpers.tryInt.

Prinz23 committed:
Add timezone correction to the Coming Episodes Page

kungktd commented:
I really hope this gets into SB. I requested this as soon as I started to use SB. 
Great work Prinz23!

simonk83 commented:
Hallelujah !!

On Fri, Mar 22, 2013 at 1:49 AM, kungktd notifications@github.com wrote:

> I really hope this gets into SB. I requested this as soon as I started to use SB. 
> 
> ## Great work Prinz23!
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/midgetspy/Sick-Beard/pull/635#issuecomment-15242789

Prinz23 commented:
> what happens if github is down? can we disable the use of the timezone?

It's displayed like now, without timezone correction. Same if the Network is not known or wrong or timezone for network is unknown, ... And network timezone info is saved in the cache.db like the scene exception in a new table. The table is only updated if needed.

Of course it's displayed on the correct day and many other fixes were currently the displaying is simply wrong.

And fixed offsets don't work. Timezones change multiple times a year!!! Most of the time DST dates change. And this code is only executed if a user clicks on the Coming Episodes link, so there is no load on the system if the user doesn't look at the Coming Episodes Page.

You can lookup how often timezones change:
ftp://ftp.iana.org/tz/releases/

...
2010: 15 times
2011: 14 times
2012: 10 times
2013: 2 times until now

Prinz23 committed:
Small bugfix: duplicate [today] Title Headers

drewfreyling commented:
+1
I also dislike the reliance on github here.

Prinz23 commented:
Scene Exceptions relies on github for years now, the exact same way. And that is a much more important function, this only affects the displaying, nothing more.

What up with the double standards?

rolo20 commented:
It works! Thanx

drewfreyling commented:
@Prinz23 Wasn't aware of the current reliance on it. I say we should go for it then and if we don't like the reliance on it we remove it everywhere not just in this instance.

Prinz23 commented:
The data has to come from somewhere. github is as good as anywhere else.

So removing is impossible or one has to remove the functions completely. Which would mean that shows with scene exceptions wouldn't be downloaded at all anymore.

simonk83 commented:
My most wanted feature for a very long time now.  Hoping for a quick pull.

On Thu, Mar 28, 2013 at 8:41 AM, Prinz23 notifications@github.com wrote:

> The data has to come from somewhere. github is as good as anywhere else.
> 
> ## So removing is impossible or one has to remove the functions completely. Which would mean that shows with scene exceptions wouldn't be downloaded at all anymore.
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/midgetspy/Sick-Beard/pull/635#issuecomment-15554952

thezoggy commented on sickbeard/helpers.py:
```diff
@@ -24,6 +24,7 @@
 import shutil
 import traceback
 import time, sys
+import hashlib
```
we do this for when we generate the md5 apikey, try the newer hashlib but fallback to the legacy md5 (pre python 2.5)

```
        try:
            from hashlib import md5
        except ImportError:
            from md5 import md5
```

Prinz23 replied:
md5 is only needed for pre python 2.5. 

Since the minimum requirement for Sickbeard to run is python 2.5, this is useless

thezoggy replied:
yes, noticed that i updated the comment from my original post as i researched it once I made the comment to see why we do it (as it was taken from sab).

Prinz23 committed:
Fixed a small error in the mass_action function

Not needed in the timezone correction pull request

Prinz23 committed:
And again a fix for the mass_action

greenythebeast commented:
Is it possible to get airtime added to the List layout of the coming episodes page?

Prinz23 committed:
Removed unneeded double checks and unneeded variables in the mass_action function

Commented out logging of each sql query in the transaction (performance)

This changed will increase the speed the transaction is executed, especially on system that have a low IO performance (removed logging which would be written to disk)

Prinz23 commented:
The Airtime is already displayed in the list layout as long the OS system setting is set to display it. It's displayed with the python datetime format function strftime in "Locale’s appropriate date and time representation."

Prinz23 committed:
Fixed error

Prinz23 committed:
activate individual queries logging as additional function parameter (Thanks @thezoggy)

Prinz23 committed:
Improved time parsing to ignore extra text at the end

mr-orange commented:
Really good work on this Prinz23. I've added this in my dev branch. Also insert 2 line of code for proper air time string formatting independent from locale setting.

Prinz23 commented:
> Also insert 2 line of code for proper air time string formatting independent from locale setting.

I didn't add this on purpose, because I live in a 24 clock country like many people and don't like (find it even confusing) 12 clocks.

I'm thinking about adding a option in config to set the format for date and time. If people want it.

arogl commented on sickbeard/network_timezones.py:
```diff
@@ -0,0 +1,166 @@
+    d = {}
+
+    # network timezones are stored on github pages
+    url = 'http://github.com/Prinz23/sb_network_timezones/raw/master/network_timezones.txt'
```
Any reason you have chosen Perth for the Australian times? Sydney would be better.

mr-orange replied:
I do not choose anything about timezone. You have to ask to Prinz23

On Mon, Apr 8, 2013 at 11:37 AM, arogl notifications@github.com wrote:

> In sickbeard/network_timezones.py:
> 
> > -            _remove_zoneinfo_failed(zonefile_tmp)
> > -            return
> > -    else:
> > -        _remove_zoneinfo_failed(zonefile_tmp)
> > -        logger.log(u"MD5 HASH doesn't match: " + zoneinfo_md5.upper() + ' File: ' + new_hash.upper(),logger.ERROR)
> > -        return
> >   +
> >   +# update the network timezone table
> >   +def update_network_dict():
> >   +
> > -    _update_zoneinfo()
> >   +
> > -    d = {}
> >   +
> > -    # network timezones are stored on github pages
> > -    url = 'http://github.com/Prinz23/sb_network_timezones/raw/master/network_timezones.txt'
> 
> Any reason you have chosen Perth for the Australian times? Sydney would be
> better.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/midgetspy/Sick-Beard/pull/635/files#r3691123
> .

Prinz23 replied:
I had to guess the correct timezone for country's with multiple timezones. Only for US and Canada I knew the correct timezone.

I have to open a Thread in the Sickbeard Forum where people can post and correct wrong timezones and request new networks too. Just like the scene exception Thread in the Forum.

Prinz23 commented:
I opened a Thread in the Forum where new or incorrect Timezones for Networks can be reported/requested. Just like the ADD Scene Exception Thread.

http://sickbeard.com/forums/viewtopic.php?f=4&t=7007

Prinz23 committed:
Added configurable Date and Time Format in Config -> General

Added Timezone correction to all Pages (Home - Coming Episodes - Show Details - Backlog Overview)
Added Warning (invalid Timeformat) to the Show Details Page if Timeformat can't be parsed
Optimized the network_timezones module a bit

ChrislyBear-GH commented:
Best fork I've used so far! Please pull it soon, cause this feature is awesome.

A note to Prinz23:
It's very nice, that you offer so many different date formats. Since I'm missing the ISO 8601 date format (yyyy-mm-dd) I have to ask if it would be possible to offer an option to directly enter a format string? 

Chris

Prinz23 committed:
Added ISO 8601 date formats like (yyyy-mm-dd)

mattymess commented:
I cloned this repo, but don't see where these changes are in 'Config > General.' Do I even need to do anything in the Sickbeard admin?

ChrislyBear-GH commented:
I believe you didn't checkout the right branch.
1. Terminate SickBeard
2. Execute `git checkout timezones-coming-eps` inside the directory in which the cloned repo resides.
3. Start SickBeard again.

Chris

mattymess commented:
I cloned Prinz23's fork so I didn't think I'd need specifically check out that branch, but I will try and let you know. Thanks for the reply, KurrKurr.

mattymess commented:
That worked like a charm. Thanks @KurrKurr!

Prinz23 committed:
Windows only accepts dates between 1970 and 2038 for localtime convert  and throws an ValueError if the date is outside

As a workaround the function now gives back the unchanged Date in case of a ValueError

Prinz23 committed:
It's better to give back a timezone aware datatime in case of an ValueError

threesquared commented:
Why was this request closed? I would really like to see this in the master branch.

simonk83 commented:
Absolutely

On Mon, May 6, 2013 at 10:30 AM, Ben Speakman notifications@github.com
wrote:

> ## Why was this request closed? I would really like to see this in the master branch.
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/midgetspy/Sick-Beard/pull/635#issuecomment-17462345

ChrislyBear-GH commented:
Yes, why the closing? It hasn't been merged yet!?

mr-orange commented:
@Prinz23 
I really liked your work on timezone integration. And it's really a pity you have closed this.
Please do not delete the branch and if you want I and other would really happy to have a pull request on my branch.
Thanks.
Mr_orange.

Prinz23 commented:
I will not delete it.

I simply will not work on anything thetvdb (for example: Sickbeard) related anymore.

Stanzilla commented:
@itofzo can this get merged, please?

Prinz23 commented:
The Pull-Request is very old and many bugfixes are missing in this very old request.

Stanzilla commented:
@Prinz23 care enough to re-do it?

Prinz23 commented:
There are many reasons why I don't bother.

And additionally: When was the last time that any pull request was merged in this Repo? (Not counting the main dev's, thezoggy)

thezoggy commented:
Stuff gets pulled, usually we pull it into our own branch to fix issues first before pushing to the repo.

-----Original Message-----
From: "Prinz23" notifications@github.com
Sent: ‎7/‎6/‎2014 12:42 PM
To: "midgetspy/Sick-Beard" Sick-Beard@noreply.github.com
Cc: "thezoggy" thezoggy@gmail.com
Subject: Re: [Sick-Beard] Add timezone correction to the Coming Episodes Page (#635)

There are many reasons why I don't bother.

And additionally: When was the last time that any pull request was merged in this Repo? (Not counting the main dev's, thezoggy)

---

Reply to this email directly or view it on GitHub:
https://github.com/midgetspy/Sick-Beard/pull/635#issuecomment-48117989

ppslim commented:
Granted, but often attribution / credit is lost in the process, although the licensing used does obviously support this :(

thezoggy commented:
Incorrect. If the pull is pulled into a branch and pushed.. The history and 'credit' is maintained.. There are things I've ported over from another branch (for example prinz nzbget changes) which we credit even though I end up rewriting or changing due to the branch differences... The credit is still there. 

-----Original Message-----
From: "Phil R" notifications@github.com
Sent: ‎7/‎6/‎2014 2:05 PM
To: "midgetspy/Sick-Beard" Sick-Beard@noreply.github.com
Cc: "thezoggy" thezoggy@gmail.com
Subject: Re: [Sick-Beard] Add timezone correction to the Coming Episodes Page (#635)

Granted, but often attribution / credit is lost in the process, although the licensing used does obviously support this :(

---

Reply to this email directly or view it on GitHub:
https://github.com/midgetspy/Sick-Beard/pull/635#issuecomment-48120303

Stanzilla commented:
Anyway, would be nice if one of you guys can redo this patch against master.

Prinz23 commented:
If at all, I would have to port the mr_orange branch (no longer existing) here, since I fixed a lot of things in that version.

Stanzilla commented:
It's kinda sad that there have to be so many branches and forks of the same thing. I guess it comes to what you said, not enough stuff getting merged from the pull requests.

thezoggy commented:
So you would rather SB be closed source? That way there wouldn't be forks or branches..

-----Original Message-----
From: "Stanzilla" notifications@github.com
Sent: ‎7/‎8/‎2014 1:25 PM
To: "midgetspy/Sick-Beard" Sick-Beard@noreply.github.com
Cc: "thezoggy" thezoggy@gmail.com
Subject: Re: [Sick-Beard] Add timezone correction to the Coming Episodes Page (#635)

It's kinda sad that there have to be so many branches and forks of the same thing. I guess it comes to what you said, not enough stuff getting merged from the pull requests. 

---

Reply to this email directly or view it on GitHub:
https://github.com/midgetspy/Sick-Beard/pull/635#issuecomment-48379930

Stanzilla commented:
Nope, just more working together with people, so more open with code audits etc.
